src/torrent.rs

```diff
@@ -108,86 +115,133 @@ impl Torrent {
-                            b.last_request = Some(Instant::now());
-                            self.requested_blocks += 1;
-                            return Some((p.index, b.offset, FIXED_BLOCK_SIZE));
-                        }
+    pub fn get_next_block(&mut self, bitfield: &BitField) -> Option<PieceIndexOffsetLength> {
+        if self.in_progress_blocks.len() == 256 {
+            // there are no more blocks for the requester to help with "right now"
+            println!(
+                "we are at capacity for new in progress blocks; current in progress: {:?}",
+                self.in_progress_blocks
+                    .iter()
+                    .map(|block| { (block.piece_index, block.offset) })
+            );
+            return None;
+        }
+
+        let res: Option<(u32, &mut VecDeque<Block>)> = {
+            let mut res = None;
+            // O(total number of pieces); always pulls pieces and blocks based on exact order of index of piece from 0 to total number of pieces
+            for piece in self.pieces.iter_mut() {
+                let piece_index = piece.index;
+
+                // relatively cheap; should not panic!!!
+                match bitfield.is_set(piece_index as usize).unwrap() {
+                    true => {
+                        let blocks_to_request_queue = &mut piece.blocks;
+                        res = Some((piece_index, blocks_to_request_queue));
+                        break;
                     }
+                    false => continue,
+                }
+            }
+            res
+        };
+
+        match res {
+            Some((piece_index, blocks_to_request_queue)) => {
+                // we can give them any block in p.index's block queue
+                let mut next_block = blocks_to_request_queue.pop_front().expect("tried to get a block from a piece's queue, but it was empty even when piece wasn't marked as done"); // It shouldn't be empty since piece was not complete...
+                let offset = next_block.offset;
+                next_block.state = BlockState::Requested;
+                next_block.last_request = Some(Instant::now());
+                self.requested_blocks += 1;
+
+                self.in_progress_blocks.push(next_block);
+
+                if blocks_to_request_queue.is_empty() {
+                    let index = self
+                        .pieces
+                        .iter()
+                        .position(|piece| piece.index == piece_index)
+                        .expect(
+                            "tried to remove a completed piece from the pieces field and failed",
+                        );
+                    self.pieces.swap_remove(index);
                 }
-            } else {
-                continue;
+
+                Some(PieceIndexOffsetLength(
+                    piece_index,
+                    offset,
+                    FIXED_BLOCK_SIZE,
+                ))
             }
+            None => None,
```

Comment:
This is a weird result to give in this case because it communicates to the caller that even though we found nothing for you to do ever, we might later.